day_one/part_one.go

```diff
@@ -1,84 +1,118 @@
 package main
 
 import (
+	"bufio"
 	_ "embed"
 	"fmt"
+	"io"
 	"log"
-	"sort"
+	"os"
 	"strconv"
 	"strings"
-)
-
-//go:embed input.txt
-var input string
 
-func init() {
-	// do this in init (not main) so test file has same input
-	input = strings.TrimRight(input, "\n")
-	if len(input) == 0 {
-		panic("empty input.txt file")
-	}
-}
+	"golang.org/x/exp/slices"
+)
 
 func main() {
-	parsedInput := parseInput(input)
+	parsedInput, err := ParseFile(os.Args[1])
+	if err != nil {
+		log.Fatalf("parse input: %v", err)
+	}
 
-	elves := groupElves(parsedInput)
+	elves := GroupElves(parsedInput)
 	for _, elf := range elves {
 		fmt.Println(elf)
 	}
 
-	elf := findElfWithMostest(elves)
+	elf := TopElf(elves)
 	fmt.Println("Elf with the mostest: ", elf)
 
-	topThreeElves := findThreeElvesWithTheMostest(elves)
-
+	topThreeElves := TopThreeElves(elves)
 	fmt.Println("Top three elves: ", topThreeElves)
 }
 
-func findThreeElvesWithTheMostest(elves []int) []int {
-	elves = sortElves(elves)
-	return elves[:3]
-}
+func TopThreeElves(elves []int) []int {
+	m := elves[:3]
+	slices.Sort(m)
 
-func parseInput(input string) []int {
-	parsed := make([]int, 0)
-	for _, row := range strings.Split(input, "\n") {
-		if row != "" {
-			i, err := strconv.Atoi(row)
-			if err != nil {
-				log.Fatal(err)
-			}
-			parsed = append(parsed, i)
-		} else {
-			parsed = append(parsed, 0)
+	for _, elf := range elves[3:] {
+		if elf > m[2] {
+			copy(m, m[1:])
+			m[2] = elf
+			continue
+		}
+
+		if elf > m[1] {
+			m[0] = m[1]
+			m[1] = elf
+			continue
+		}
+
+		if elf > m[0] {
+			m[0] = elf
 		}
 	}
-	return parsed
+	m[0], m[2] = m[2], m[0]
+
+	return m
 }
 
-func findElfWithMostest(elves []int) int {
-	elves = sortElves(elves)
-	return elves[0]
+func ParseFile(path string) ([]int, error) {
```

Comment:
One thing that has been bugging me each day is the initial parsing. How would we go about having a `ParseFile` function that reads the file, but also takes another function(not sure if that is done in Go?) to help set the output the way the `Parse` function works here.

Reply:
I don't quite understand what you're asking. Higher order functions are quite possible in Go, i.e. `func Example(val string, f func(rune) bool) error { ... }`, but I'm not sure what it is you want to do with one in this case.

Reply:
Another thing I was doing in Python was to stream the input line by line by creating a generator and never having the whole file in memory. Is there a similar way to do that in Go? 

eg: open the file, read the first line, parse it, send it off to PartOne or similar?

Reply:
> I don't quite understand what you're asking. Higher order functions are quite possible in Go, i.e. `func Example(val string, f func(rune) bool) error { ... }`, but I'm not sure what it is you want to do with one in this case.

All good, knowing its possible is enough. Ill have a crack at what I was thinking for the next day (lets see if I can get some free time in the next few days :-\)

Thanks for your input on this, it's truly helpful!